crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -3229,6 +3300,85 @@ pub(crate) mod tests {
         );
     }
 
+    /// Check that register_operation correctly registers operations from MQTT and handles error conditions
+    #[tokio::test]
+    async fn test_register_entity() {
+        let tmp_dir = TempTedgeDir::new();
+        let mut config = c8y_converter_config(&tmp_dir);
+        config.enable_auto_register = false;
+
+        let (mut converter, _http_proxy) = create_c8y_converter_from_config(config);
+
+        // main device command
+        let main_device = converter.entity_store.main_device().clone();
+
+        let operation_topic = converter.mqtt_schema.topic_for(
+            &main_device,
+            &Channel::CommandMetadata {
+                operation: OperationType::Custom("my_operation".to_string()),
+            },
+        );
+        let operation_msg = MqttMessage::new(&operation_topic, "{}");
+
+        converter.convert(&operation_msg).await;
+
+        // child device command
+        let reg_message = MqttMessage::new(
+            &Topic::new_unchecked("te/device/child0//"),
+            json!({
+                "@type": "child-device",
+                "@id": "child0",
+                "name": "child0",
+                "@parent": "device/main//",
+            })
+            .to_string(),
+        );
+
+        converter
+            .try_register_source_entities(&reg_message)
+            .await
+            .unwrap();
+
+        let child_device = EntityTopicId::default_child_device("child0").unwrap();
+        let operation_topic = converter.mqtt_schema.topic_for(
+            &child_device,
+            &Channel::CommandMetadata {
+                operation: OperationType::Custom("my_operation".to_string()),
+            },
+        );
+        let operation_msg = MqttMessage::new(&operation_topic, "{}");
+
+        converter.convert(&operation_msg).await;
+
+        // service command
+        let reg_message = MqttMessage::new(
+            &Topic::new_unchecked("te/device/main/service/service0"),
+            json!({
+                "@type": "service",
+                "@id": "service0",
+                "name": "service0",
+                "@parent": "device/main//",
+            })
+            .to_string(),
+        );
+
+        converter
+            .try_register_source_entities(&reg_message)
+            .await
+            .unwrap();
+
+        let service = main_device.default_service_for_device("service0").unwrap();
+        let operation_topic = converter.mqtt_schema.topic_for(
+            &service,
+            &Channel::CommandMetadata {
+                operation: OperationType::Custom("my_operation".to_string()),
+            },
+        );
+        let operation_msg = MqttMessage::new(&operation_topic, "{}");
+
+        converter.convert(&operation_msg).await;
```

Comment:
Even though we're not expecting any 114 messages for this service case, it'd be better to do an explicit assertion that the returned vector is empty.

Reply:
Turned out the vector is not empty, there's an MQTT message with an error about operations for services being unsupported

fixed in 01ec205849